worker/src/socket.rs

```diff
@@ -0,0 +1,432 @@
+enum Closing {
+    Pending(JsFuture),
+    None,
+}
+
+impl Default for Closing {
+    fn default() -> Self {
+        Self::None
+    }
+}
+
+/// Represents an outbound TCP connection from your Worker.
+pub struct Socket {
+    inner: worker_sys::Socket,
+    writable: web_sys::WritableStream,
+    readable: web_sys::ReadableStream,
+    write: Option<Writing>,
+    read: Option<Reading>,
+    close: Option<Closing>,
+}
+
+// This can only be done because workers are single threaded.
+unsafe impl Send for Socket {}
+unsafe impl Sync for Socket {}
+
+impl Socket {
+    fn new(inner: worker_sys::Socket) -> Self {
+        let writable = inner.writable();
+        let readable = inner.readable();
+        Socket {
+            inner,
+            writable,
+            readable,
+            read: None,
+            write: None,
+            close: None,
+        }
+    }
+
+    /// Closes the TCP socket. Both the readable and writable streams are forcibly closed.
+    pub async fn close(&mut self) -> Result<()> {
+        JsFuture::from(self.inner.close()).await?;
+        Ok(())
+    }
+
+    /// This Future is resolved when the socket is closed
+    /// and is rejected if the socket encounters an error.
+    pub async fn closed(&self) -> Result<()> {
+        JsFuture::from(self.inner.closed()).await?;
+        Ok(())
+    }
+
+    /// Upgrades an insecure socket to a secure one that uses TLS,
+    /// returning a new Socket. Note that in order to call this method,
+    /// you must set [`secure_transport`](SocketOptions::secure_transport)
+    /// to [`StartTls`](SecureTransport::StartTls) when initially
+    /// calling [`connect`](connect) to create the socket.
+    pub fn start_tls(self) -> Socket {
+        let inner = self.inner.start_tls();
+        Socket::new(inner)
+    }
+
+    pub fn builder() -> ConnectionBuilder {
+        ConnectionBuilder::default()
+    }
+
+    fn handle_write_future(
+        cx: &mut std::task::Context<'_>,
+        mut fut: JsFuture,
+        writer: web_sys::WritableStreamDefaultWriter,
+        len: usize,
+    ) -> (Writing, Poll<std::io::Result<usize>>) {
+        match fut.poll_unpin(cx) {
+            Poll::Pending => (Writing::Pending(fut, writer, len), Poll::Pending),
+            Poll::Ready(res) => {
+                writer.release_lock();
+                match res {
+                    Ok(_) => (Writing::None, Poll::Ready(Ok(len))),
+                    Err(e) => (Writing::None, Poll::Ready(Err(js_value_to_std_io_error(e)))),
+                }
+            }
+        }
+    }
+}
+
+fn js_value_to_std_io_error(value: JsValue) -> std::io::Error {
+    let s = if value.is_string() {
+        value.as_string().unwrap()
+    } else if let Some(value) = value.dyn_ref::<js_sys::Error>() {
+        value.to_string().into()
+    } else {
+        format!("Error interpreting JsError: {:?}", value)
+    };
+    std::io::Error::new(std::io::ErrorKind::Other, s)
+}
+impl tokio::io::AsyncRead for Socket {
```

Comment:
same nit here, I think a lot of these should use imports instead of qualified paths. eg:

```rust
use std::io::Result as IOResult;
use std::pin::Pin;
use std::task::{Context, Poll};

use tokio::io::{ReadBuf, AsyncRead};

// ....

impl AsyncRead for Socket {
    fn poll_read(
        mut self: Pin<&mut Self>,
        cx: &mut Context<'_>,
        buf: &mut ReadBuf<'_>,
    ) -> Poll<IOResult<()>> {
        // ...
    }
```